node/pod.go

```diff
@@ -216,9 +216,29 @@ func (pc *PodController) updatePodStatus(ctx context.Context, podFromKubernetes
 	kPod.Lock()
 	podFromProvider := kPod.lastPodStatusReceivedFromProvider.DeepCopy()
 	kPod.Unlock()
-	if cmp.Equal(podFromKubernetes.Status, podFromProvider.Status) {
+	if cmp.Equal(podFromKubernetes.Status, podFromProvider.Status) && podFromProvider.DeletionTimestamp == nil {
 		return nil
 	}
+	// Pod deleted by provider due some reasons. e.g. a K8s provider, pod created by deployment would be evicted when node is not ready.
+	// If we do not delete pod in K8s, deployment would not create a new one.
+	if podFromProvider.DeletionTimestamp != nil && podFromKubernetes.DeletionTimestamp == nil {
+		deleteOptions := metav1.DeleteOptions{
+			GracePeriodSeconds: podFromProvider.DeletionGracePeriodSeconds,
+		}
+		if podFromProvider.DeletionTimestamp.After(time.Now()) {
+			deleteOptions.GracePeriodSeconds = new(int64)
+		}
+		// check status here to avoid pod re-created deleted incorrectly. e.g. delete a pod from K8s and re-create it(statefulSet and so on),
+		// pod in provider may not delete immediately. so deletionTimestamp is not nil. Then the re-created one would be deleted if we do not check pod status.
+		if cmp.Equal(podFromKubernetes.Status, podFromProvider.Status) {
+			if err := pc.client.Pods(podFromKubernetes.Namespace).Delete(ctx, podFromKubernetes.Name, deleteOptions); err != nil && !errors.IsNotFound(err) {
```

Comment:
To 1 and 2:

The original logic is if status is equal we return  directly, right? This PR add another logic: this pod in provider should not be deleing.

so here, we just judge if the pod in provider is deleting. If the pod is deleting, we delete the pod in K8s.

To 3.

update will not be dropped if you do not delete it in provider.

Reply:
@cwdsuzhou Right, but there can be the case where you delete the pod in the provider, but you still want to do status updates indicating that you're tearing down all of the containers in the pod. For example, you may want to delete the pod, and then set the container status to not running to indicate to users that certain containers are successfully being torn down in the pod.

Reply:
I wonder whether we still need update pod status in K8s since the pod in provider is under deleting. I do not think it’s meaningful to do this.

Reply:
I think you can take this PR and have a try.

Reply:
So, checking it -- you miss one update, and then subsequent updates work.

Reply:
Can you write a test here which deletes the pod from the provide and marks all the containers as exited and I think the PR will be good to go.

Reply:
Base on the current test framework, it may be hard to be tested through unit tests. All updateStatus funcs are not tested now.

there 2 scene:
1. delete pod and mark as exited.

IMO this would not happen. How can you delete a pod and update container status in one request?  If this happens, do you think this a `delete` operation? Why not separate it int two?

2. delete pod, then mark as exited.

pod in K8s would then be deleting. When container exited, pod status in K8s and provider are different, so status in K8s would happen.

Reply:
@sargun  some simple unit tests added.  PTAL

Reply:
I think if you drop the "return nil" on line 240, it'll achieve what I'm suggesting.

Reply:
If the provider does NotifyPods(Pod), where Pod has BOTH delete timestamp set, wouldn't you want both values to make it to API server? Specifically, wouldn't you want to do the UpdateStatus(...) first, then the delete -- also, a question, as part of updateStatus, can you set DeletionTimestamp / DeletionGracePeriod?

Reply:
```
I think if you drop the "return nil" on line 240, it'll achieve what I'm suggesting.
```

Since pod deleted success, why not return nil?

Reply:
```
If the provider does NotifyPods(Pod), where Pod has BOTH delete timestamp set, wouldn't you want both values to make it to API server? Specifically, wouldn't you want to do the UpdateStatus(...) first, then the delete -- also, a question, as part of updateStatus, can you set DeletionTimestamp / DeletionGracePeriod?
```

You can have a try,  set `DeletionTimestamp`? the value would not update in a real K8s not fake. Also, unit tests have added, the logic newly add not break anything.

Reply:
@cpuguy83 PTAL

Reply:
1. We continue to log the "update"
2. If the container statuses *also* changed and were part of the same notify, wouldn't we want to call UpdateStatus?

Rather than require the provider to do two notifies?

Reply:
> I think if you drop the "return nil" on line 240, it'll achieve what I'm suggesting.

Dropped

Reply:
> If the provider does NotifyPods(Pod), where Pod has BOTH delete timestamp set, wouldn't you want both values to make it to API server? Specifically, wouldn't you want to do the UpdateStatus(...) first, then the delete -- also, a question, as part of updateStatus, can you set DeletionTimestamp / DeletionGracePeriod?

You can have a try,  set `DeletionTimestamp`? the value would not update in a real K8s not fake. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#L162


Also, unit tests have added, the logic newly add not break anything.

Reply:
I like the unit tests. Thank you.

Reply:
> 1. We continue to log the "update"
> 2. If the container statuses _also_ changed and were part of the same notify, wouldn't we want to call UpdateStatus?
> 
> Rather than require the provider to do two notifies?

Get it. Updated

PS: if  provider is also a K8s cluster. this would not happen